### DISCUSSION TITLE
How to deal with component with async and sync possbilities?

### DISCUSSION THREAD

**Discussion by zh0uquan** [DavidVujic/python-polylith#279] [Q&A] [answered]
Lets say I have a component have both sync and async ability, what is the best practice to organize this component, should I separate them into two component? like `sync_create` and `async_create`? and how about the dependency management? since both of third party deps will stay in the root `pyproject.toml`, is there any concern about this?

**Comment by DavidVujic** [comment, marked as answer]
If they do very different things (and use different third-party dependencies) behind the functions, then it could be good to separate them into different components. If it is the same basic functionality in both of them, it could be a good thing to have them in the same component. Or even extract some of the shared code into separate component(s) too (smaller LEGO bricks).

In short, it depends 😄 (sorry)

If you keep the components small, then it shouldn't be too much work changing them to separate/combined if it turns out it was the wrong decision.

**Reply by zh0uquan** [reply]
Thanks for the answers! 👍 
Currently we are applying the polylith monorepo in our team, and we are adjusting our own IaC and versioning system in our CI/CD to adapt the polylith philosophy. We find CI/CD docs are kinda missing in the documentation, where I think there might be some best practice out there, eg: how to manage projects versions (do we do VCS or pyproject.toml). Do you have to plan to extend the docs in the future?
In general we are happy, we find our common code and projects are better managed,

**Reply by DavidVujic** [reply]
I'm happy to hear that you like Polylith!

Yes, the CI parts isn't covered in the docs and something that I probably should add.

I wrote a [blog post about Kubernetes](https://davidvujic.blogspot.com/2024/08/simple-kubernetes-in-python-monorepos.html), and how you can set it up in a similar way as with the Python code.

Manage project versions: I don't think it is different from a single-project repo (maybe I misunderstand)? I usually bump the version for individual projects with semantic versioning, but that is outside of Polylith. The repo itself, I would recommend to use git tags. With tags, you can run diffs using the poly tool.
